### PR TITLE
Added timestamp and location to LOG messages

### PR DESCRIPTION
**Pull request by ChrisCheng816**
Added timestamps and locations to log messages to address #886.

### CODE REVIEW

**Review by rafaelsamenezes — CHANGES_REQUESTED** [review]
Thanks for the PR. It looks like a good direction, could you:

1. Check if its possible to just preapend at the format string in https://github.com/esbmc/esbmc/blob/master/src/util/message.h#L73 that way you will not need to do as many changes. 
2. As suggested at https://github.com/esbmc/esbmc/issues/886#issuecomment-1463333671 we need this to be a flag or env variable opt-in.

**Comment by mikhailramalho** [comment]
What does the log look like after your change?

**Comment by ChrisCheng816** [comment]
> Thanks for the PR. It looks like a good direction, could you:
> 
> 1. Check if its possible to just preapend at the format string in https://github.com/esbmc/esbmc/blob/master/src/util/message.h#L73 that way you will not need to do as many changes.
> 2. As suggested at [Add timestamp and location to LOG messages #886 (comment)](https://github.com/esbmc/esbmc/issues/886#issuecomment-1463333671) we need this to be a flag or env variable opt-in.

1. Yes, I've already made all the changes in message.h.
2. I've opted it in as a flag, and users can use `--log-message` as an option.

**Comment by ChrisCheng816** [comment]
> What does the log look like after your change?

For example, if the user turns on the option `--log-message`, the log will look like `VERIFICATION SUCCESSFUL File: /ESBMC_Project/esbmc/src/esbmc/bmc.cpp, Line: 210 [2024-10-29 16:30:21] ` when verification is successful.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 1]
```diff
@@ -490,6 +490,11 @@ void esbmc_parseoptionst::get_command_line_options(optionst &options)
   if (cmdline.isset("cvc"))
     options.set_option("cvc4", true);
 
+  if (cmdline.isset("log-message"))
+  {
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: no need for curly brackets here.

File: src/esbmc/esbmc_parseoptions.cpp [hunk 2]
```diff
@@ -2272,10 +2279,12 @@ static void collect_symbol_names(
...
-      if (!is_nil_expr(e))
-        collect_symbol_names(e, prefix, used_syms);
-    });
+    e->foreach_operand(
+      [&prefix, &used_syms](const expr2tc &e)
+      {
+        if (!is_nil_expr(e))
+          collect_symbol_names(e, prefix, used_syms);
+      });
   }
```
Comment (mikhailramalho) [inline src/esbmc/esbmc_parseoptions.cpp]: is this just code style? We have to check if it's the same version used in our CI.

**Review by mikhailramalho — CHANGES_REQUESTED** [review]
I don't like the time being appended and would prefer it to be prepended. This is more the standard in other tools, like journalctl, dmesg, etc.

Let's hear from the other developers about the format before landing this.

**Review by rafaelsamenezes — APPROVED** [review]
I also prefer the time to be preappended. Otherwise lgtm.

**Review by lucasccordeiro — APPROVED** [review]
I also agree with @mikhailramalho's suggestions. Please address them before we merge this PR, @Chriszai.

**Comment by lucasccordeiro** [comment]
@Chriszai: can you also please fix the code style issue here?

**Comment by ChrisCheng816** [comment]
> @Chriszai: can you also please fix the code style issue here?

Yes, sure

**Comment by lucasccordeiro** [comment]
CI is still failing, @Chriszai.

**Comment by lucasccordeiro** [comment]
CI is still failing, @Chriszai.

**Comment by lucasccordeiro** [comment]
image

**Comment by lucasccordeiro** [comment]
````
error: Your local changes to the following files would be overwritten by checkout:
	src/esbmc/options.cpp
	src/util/message.h
Please commit your changes or stash them before you switch branches.
Aborting
Error: Invalid status code: 1
    at ChildProcess.<anonymous> (/home/runner/work/_actions/stefanzweifel/git-auto-commit-action/v5/index.js:17:19)
    at ChildProcess.emit (node:events:519:28)
    at maybeClose (node:internal/child_process:1105:16)
    at ChildProcess._handle.onexit (node:internal/child_process:305:5) {
  code: 1
}
Error: Invalid status code: 1
    at ChildProcess.<anonymous> (/home/runner/work/_actions/stefanzweifel/git-auto-commit-action/v5/index.js:17:19)
    at ChildProcess.emit (node:events:519:28)
    at maybeClose (node:internal/child_process:1105:16)
    at ChildProcess._handle.onexit (node:internal/child_process:305:5)
````

**Comment by ChrisCheng816** [comment]
I guess it has nothing to do with my code, but with the git workflow

**Comment by lucasccordeiro** [comment]
Thanks, @Chriszai.

@mikhailramalho: do you still have further comments about this PR?

File: src/util/message.h [hunk 3]
```diff
@@ -75,8 +75,12 @@ struct messaget
...
       {
-        std::time_t currentTime = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
-        std::string timeStr = (std::ostringstream{} << std::put_time(std::localtime(&currentTime), "%Y-%m-%d %H:%M:%S")).str();
+        std::time_t currentTime = std::chrono::system_clock::to_time_t(
+          std::chrono::system_clock::now());
+        std::string timeStr =
+          (std::ostringstream{}
+           << std::put_time(std::localtime(&currentTime), "%Y-%m-%d %H:%M:%S"))
+            .str();
         fmt::print(f, "[{}] File: {}, Line: {} ", timeStr, file, line);
```
Comment (lucasccordeiro) [inline src/util/message.h]: @Chriszai: According to @mikhailramalho's comments, should you move `timeStr` to the beginning, or did I miss something here?

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]

File: src/util/message.h (hunk 3 above)
Comment (ChrisCheng816) [reply src/util/message.h]: Yes, I have moved `timeStr` to the beginning; during the verification process, timestamp will be output first (as shown in the line 84), followed by LOG messages (as shown in the line 86).

```
fmt::print(f, "[{}] File: {}, Line: {} ", timeStr, file, line);
println(f, lvl, format, fmt::make_format_args(args...));
```

**Comment by lucasccordeiro** [comment]
Thanks for this explanation!

**Comment by lucasccordeiro** [comment]
image

**Comment by rafaelsamenezes** [comment]
@Chriszai could you please rebase your code on top of master (we have a guide here: https://github.com/esbmc/esbmc/blob/master/CONTRIBUTIONS.md#contributing-to-the-code-base). After that we should be able to merge.

**Review by mikhailramalho — APPROVED** [review]
LGTM

**Comment by ChrisCheng816** [comment]
image
My fork is already updated; still cannot be merged?

**Comment by lucasccordeiro** [comment]
@Chriszai: We're still unable to merge your PR.

Did you try to follow the steps indicated by @rafaelsamenezes?

**Pull request closed without merging**